File: ts/test/perf.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */

import { test, describe } from 'node:test'
import assert from 'node:assert'

import { Tabnas } from '@tabnas/parser'
import { jsonic } from '@tabnas/jsonic'
import { MultiSource, MultiSourceOptions } from '../dist/multisource'
import { makeMemResolver } from '../dist/resolver/mem'

// Unlike sibling grammar plugins (@tabnas/yaml, @tabnas/json), multisource
// exposes NO package-level convenience `parse`: it is a plugin that callers
// install themselves via `new Tabnas().use(jsonic).use(MultiSource, opts)`.
// There is therefore no default instance to cache. The performance risk it
// shares with those plugins is the same one regardless: building the engine +
// multisource grammar dominates a parse, so rebuilding the instance on every
// parse is many times slower than building it once and reusing it.
//
// This test guards that representative usage. It compares N parses that
// rebuild the instance each time against N parses that reuse ONE instance, on
// the SAME machine in the SAME run, and asserts reuse is not slower (and is in
// fact much faster). The comparison is machine-INDEPENDENT — both sides scale
// together on a slow CI box — so there is deliberately NO wall-clock budget.
//
// If the multisource setup ever regressed to be rebuilt per parse inside a
// hot loop, the reuse path would lose its advantage and this guard would flag
// it. The test also documents the intended fast pattern: build once, reuse.
describe('perf', () => {
  test('reuse-one-instance-is-fast', () => {
    const opts: MultiSourceOptions = {
      resolver: makeMemResolver({
        'a.jsonic': 'a:1',
        'b.jsonic': 'b:2',
      }),
    }
    const src = '{x:@a.jsonic,y:@b.jsonic,z:3}'
    const n = 400

    const build = () => new Tabnas().use(jsonic).use(MultiSource, opts)

    // Warm both paths so the comparison is steady-state.
    const warm = build()
    for (let i = 0; i < 100; i++) {
      warm.parse(src)
    }
    for (let i = 0; i < 100; i++) {
      build().parse(src)
    }

    // Rebuild-per-parse: the slow anti-pattern.
    const t0 = process.hrtime.bigint()
    for (let i = 0; i < n; i++) {
      build().parse(src)
    }
    const rebuild = Number(process.hrtime.bigint() - t0)

    // Reuse one instance: the fast, intended pattern.
    const j = build()
    const t1 = process.hrtime.bigint()
    for (let i = 0; i < n; i++) {
      j.parse(src)
    }
    const reuse = Number(process.hrtime.bigint() - t1)

    const ratio = rebuild / reuse

    // Sanity: reuse produces correct results across the run.
    assert.deepEqual(j.parse(src), { x: { a: 1 }, y: { b: 2 }, z: 3 })

    // Reusing one instance must be at least as fast as rebuilding each time
    // (it is many times faster in practice — building the grammar dominates).
    // Allow 1.5x reuse-over-rebuild slack for scheduling noise; a rebuild-per
    // -parse loop is far above 1x, so this catches the regression without any
    // absolute wall-clock dependency.
    assert.ok(
      reuse < rebuild * 1.5,
      `reusing one instance should not be slower than rebuilding per parse: ` +
        `rebuild=${(rebuild / 1e6).toFixed(1)}ms reuse=${(reuse / 1e6).toFixed(1)}ms ` +
        `(rebuild/reuse ratio ${ratio.toFixed(2)}x). Build the engine once and reuse it.`,
    )

    // Surface the observed speedup for the record.
    console.log(
      `perf: rebuild=${(rebuild / 1e6).toFixed(1)}ms reuse=${(reuse / 1e6).toFixed(1)}ms ` +
        `speedup=${ratio.toFixed(2)}x`,
    )
  })
})

File: rs/tests/perf_test.rs
// The performance trap this plugin exposes, pinned the way
// ts/test/perf.test.ts and go/perf_test.go pin it: building the instance
// (the engine, the jsonic grammar, the directive and this plugin's own
// alts) dominates a small parse, so a caller must build ONE instance and
// reuse it.
//
// The check is machine-INDEPENDENT: it compares rebuild-per-call against
// instance reuse on the SAME machine in the SAME run, so a slow box
// cannot make it flaky (both sides scale together). There is
// deliberately NO wall-clock budget.
//
// It is also ONE test, deliberately. The harness runs a binary's tests in
// parallel unless told otherwise, and `ci/rust/run.sh` runs a plain
// `cargo test --all-targets` with no `--test-threads`, so splitting the
// two measurements would let one test's expensive rebuild loop overlap
// the other's cheap reuse loop and make the ratio a function of core
// count rather than of the code.

use std::time::{Duration, Instant};

use tabnas::Tabnas;
use tabnas_multisource::{make_with, MapResolver, MultiSourceOptions};

const SRC: &str = "{x:@a.jsonic,y:@b.jsonic,z:3}";
const N: usize = 120;

/// The ratio reuse has to beat. Rebuilding installs jsonic, the
/// directive and this plugin's grammar, so the real margin is far
/// larger; 4x is the floor a regression has to stay above.
const WANT: u32 = 4;

fn build() -> Tabnas {
    make_with(MultiSourceOptions::new(MapResolver::from([
        ("a.jsonic", "a:1"),
        ("b.jsonic", "b:2"),
    ])))
}

/// `N` parses through `parse_once`, timed. Every measurement goes
/// through here so the loops are identical apart from what they call.
fn time_parses(mut parse_once: impl FnMut()) -> Duration {
    let started = Instant::now();
    for _ in 0..N {
        parse_once();
    }
    started.elapsed()
}

#[test]
fn reusing_one_instance_is_far_cheaper_than_rebuilding_per_parse() {
    // Warm both paths so the comparison is steady-state.
    for _ in 0..20 {
        build().parse(SRC).expect("parses");
    }
    let shared = build();
    for _ in 0..50 {
        shared.parse(SRC).expect("parses");
    }

    // The anti-pattern: rebuild the instance on every parse.
    let rebuild = time_parses(|| {
        build().parse(SRC).expect("rebuild parse");
    });

    // The intended pattern: build once, reuse for every parse.
    let reuse = time_parses(|| {
        shared.parse(SRC).expect("reuse parse");
    });

    // Guard the baseline before dividing by it. If the rebuild loop
    // measured as no time at all then nothing was measured, and the
    // ratio below would be vacuously true.
    assert!(
        rebuild > Duration::ZERO,
        "{N} rebuild-per-call parses measured as zero elapsed time: this is \
         testing the clock, not the code"
    );

    let ratio = rebuild.as_secs_f64() / reuse.as_secs_f64().max(f64::EPSILON);
    eprintln!("perf: rebuild-per-call={rebuild:?} reuse={reuse:?} ratio={ratio:.2}x");

    assert!(
        rebuild >= WANT * reuse,
        "reusing one instance is not meaningfully faster than rebuilding per parse: \
         {N} rebuild-per-call parses took {rebuild:?} vs {reuse:?} reusing one instance \
         (ratio {ratio:.1}x, want >={WANT}x). Build one make_with instance and reuse it."
    );
}
